en/docs/learn/fraud-detection.md

```diff
@@ -0,0 +1,422 @@
+[AbstractEventHandler](https://github.com/wso2/carbon-identity-framework/blob/master/components/identity-event/org
+.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/handler/AbstractEventHandler.java) and
+override its handleEvent method. We can easily write a custom data publisher extending the
+AbstractEventHandler and overriding the handleEvent method for the purpose of publishing authentication/session data
+to any third party platform that facilitates fraud detection. When the event publisher triggers an event, there is a
+dedicated service that will transfer this event from the Identity Server to the corresponding destination. This
+service is handled by the Event Adapters. Event adaptors bridge the connection between the data source and the data
+sink. There are several output event adapters supported by the WSO2 products such as **HTTP**, **JMS**, **Web Socket**,
+**SMTP** etc.
+
+Following is a sample code segment on writing a custom data publisher.
+
+
+``` java
+public class SampleEventPublisher extends AbstractEventHandler {
+
+    @Override
+    public String getName() {
+
+        return "SampleEventPublisher";
+    }
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
+
+        if (IdentityEventConstants.EventName.AUTHENTICATION_STEP_SUCCESS.name().equals(event.getEventName()) ||
+                IdentityEventConstants.EventName.AUTHENTICATION_STEP_FAILURE.name().equals(event.getEventName())) {
+            AuthenticationData authenticationData = AnalyticsLoginDataPublisherUtils.buildAuthnDataForAuthnStep(event);
+            publishAuthenticationData(authenticationData);
+        } else if (IdentityEventConstants.EventName.AUTHENTICATION_SUCCESS.name().equals(event.getEventName()) ||
+                IdentityEventConstants.EventName.AUTHENTICATION_FAILURE.name().equals(event.getEventName())) {
+            AuthenticationData authenticationData = AnalyticsLoginDataPublisherUtils.
+                    buildAuthnDataForAuthentication(event);
+            publishAuthenticationData(authenticationData);
+        } else {
+            log.error("Event " + event.getEventName() + " cannot be handled");
+        }
+    }
+}
+```
+
+Detailed information on writing and configuring event publishers can be found [here](https://medium
+.com/identity-beyond-borders/writing-a-custom-event-publisher-for-wso2-is-5-8-0-6558d9130dde).
+
+
+### Intercepting the Authentication Flow Based on Service Response
+
+#### Custom Adaptive Authentication Script
+
+With the [adaptive authentication feature](https://is.docs.wso2.com/en/5.10.0/learn/adaptive-authentication/) in WSO2
+ Identity Server, it is possible to dynamically alter the authentication sequence with JavaScript based
+ authentication schemes. We have some predefined javascript functions to perform authenticated user related
+ operations. We also have the flexibility to plug in custom functions for desired use-cases. Similarly we can
```

Comment:
WSO2 IS also allows you the flexibility of plugging in custom functions for desired use cases. This means that you can implement